### PR TITLE
oop-object_independence-hard-q4

### PR DESCRIPTION
**Pull request by SamamthaH**
## PR Checklist
- [x]  Are helper methods really needed here?
    > Opt for hardcoding input/output into the input provider when possible
- [x]  Are the assertion fail messages good?
    - [x]  Logical (does it make sense?)
    - [x]  Readable (can a student understand what’s wrong?)
    - [x]  Not too feed-y (does it give away the answer?)
    - [x]  Grammatically sound (is the grammar and punctuation in the sentence proper?)
    - [x]  Specific (does the fail message relate to the question?)
- [x]  Is the code formatted well?
    > Run `CTRL` + `ALT` + `L` in IntelliJ
- [x]  EOF newlines
- [x]  Do the comments for Parsons follow the correct convention?
- [x]  Is the question being tested sufficiently?
    - [x]  Are there representative cases to test the program?
    - [x]  What about edge cases? (empty strings, negatives, infinity, off-by-one errors, etc)
        > If the question doesn’t specify what to do, mark it as a bad question
- [x]  Are the test method names good?
    > Test method names show up in a “What went well” section on the website, so “mathTest” will become “What went well: math test”, which doesn’t make much sense. A better name in this example would be “calculatesTaxCorrectly” so it shows up as “What went well: calculates tax correctly”

### CODE REVIEW

File: src/oop/object_independence/hard/q4/MainTest.java [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+    private final String setStringAttributeName1 = "setName";
+    private final String setStringAttributeName2 = "setQuality";
+    private final String testStringAttributeName1a = "The Starry Night";
+    private final String testStringAttributeName2a = "The Scream";
+    private final String testStringAttributeName1b = "good";
+    private final String testStringAttributeName2b = "poor";
+    private final String testStringAttributeNameUpdate = "fair";
+    private ObjectTest classInstance;
+    private ObjectTest testInstance;
+
```
Comment (KaiKhademi) [inline src/oop/object_independence/hard/q4/MainTest.java:36]: Refer to the previous PR's of this category.

**Review by KaiKhademi — CHANGES_REQUESTED** [review]

File: src/oop/object_independence/hard/q4/MainTest.java [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+        assertTrue(classInstance.hasModifier(stringAttributeName2, "private"),
+                String.format(incorrectVisibilityModifier, objectClassName, stringAttributeName2));
+    }
+
+    @Test
+    public void paintingClassHasRequiredConstructor() {
+        Class<?>[] arguments = {String.class, String.class};
+        assertTrue(classInstance.hasConstructor(arguments),
+                String.format("Your %s class is missing a required constructor.", testClassName));
+        assertTrue(classInstance.hasConstructor(arguments, new String[]{"public"}),
```
Comment (KaiKhademi) [inline src/oop/object_independence/hard/q4/MainTest.java:64]: Refer to discord

File: src/oop/object_independence/hard/q4/MainTest.java [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+
+    @ParameterizedTest
+    @MethodSource("constructorInputProvider")
+    public void paintingConstructorInitializesAttributesCorrectly(String n, String q) throws Throwable {
+        Object[][] arguments = {
+                {n, String.class},
+                {q, String.class}
+        };
+        String incorrectVisibilityModifier = "Your %s constructor has the incorrect visibility modifier.";
+        assertTrue(classInstance.hasModifier(new Class[]{String.class, String.class}, "public"),
```
Comment (KaiKhademi) [inline src/oop/object_independence/hard/q4/MainTest.java:85]: You do not need this here

**Review by KaiKhademi — CHANGES_REQUESTED** [review]

**Review by KaiKhademi — DISMISSED** [review]

File: src/oop/object_independence/hard/q4/QuestionText.md [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+- Constructor:
+    - A constructor visible from another class which accepts the parameters `String, String` and sets the **name**
+      and **quality** attributes with the passed values.
+    - This constructor should be visible to other classes
+
+- Methods (all the following methods should be visible from another class):
+    - A method called `getName` that returns the value of the **name** attribute
+    - A method called `getQuality` that returns the value of the **quality** attribute
+    - A method called `setName` that accepts a `String` as a parameter, updates the **name** attribute to the passed
```
Comment (Hedgemon4) [inline src/oop/object_independence/hard/q4/QuestionText.md:28]: Please add the return type for the set methods.

File: src/oop/object_independence/hard/q4/MainTest.java [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+        Object instance = classInstance.createInstance(arguments);
+        String incorrectGetMethods = "Your %s method does not return the value of the %s attribute.";
+
+        Object output = classInstance.callMethod(getStringAttributeName1, new String[]{"public"}, instance);
+        _assertEquals(name, output, String.format(incorrectGetMethods, getStringAttributeName1, stringAttributeName1));
+    }
+
+    @ParameterizedTest
+    @MethodSource("constructorInputProvider")
+    public void paintingHasCorrectGetQualityMethod(String name, String quality) throws Throwable {
```
Comment (Hedgemon4) [inline src/oop/object_independence/hard/q4/MainTest.java:114]: Please check the existence of get and set methods before calling them.

File: src/oop/object_independence/hard/q4/MainTest.java [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+    @ParameterizedTest
+    @MethodSource("constructorInputProvider")
+    public void paintingClassHasCorrectToStringMethod(String name, String quality) throws Throwable {
+        Object[][] arguments = {
+                {name, String.class},
+                {quality, String.class}
+        };
+        Object instance = classInstance.createInstance(arguments);
+        String expectedOutput = "The item " + name + " is of " + quality + " quality";
+        Object output = classInstance.callMethod("toString", new String[]{"public"}, instance);
```
Comment (Hedgemon4) [inline src/oop/object_independence/hard/q4/MainTest.java:180]: Also, check the existence of the toString method before calling it.

File: src/oop/object_independence/hard/q4/MainTest.java [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+                String.format(incorrectSetGet, getStringAttributeName1, stringAttributeName1, setStringAttributeName1));
+        _assertEquals(newQuality, output2,
+                String.format(incorrectSetGet, getStringAttributeName2, stringAttributeName2, setStringAttributeName2));
+        _assertEquals(expectedToStringOutput, toStringOutput,
+                String.format(incorrectToString, stringAttributeName1, stringAttributeName2,
+                        setStringAttributeName1, setStringAttributeName2));
+    }
+
+    @Test
+    public void testMuseumMainMethodProducesCorrectOutput() throws Throwable {
```
Comment (Hedgemon4) [inline src/oop/object_independence/hard/q4/MainTest.java:213]: Please add Tags to show that this test is dependent on others

Comment (Hedgemon4) [reply src/oop/object_independence/hard/q4/MainTest.java:213]: See #664 for an example

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

File: src/oop/object_independence/hard/q4/QuestionText.md (hunk 4 above)
Comment (SamamthaH) [reply src/oop/object_independence/hard/q4/QuestionText.md:28]: It already does? Unless "and returns nothing" isn't clear?

**Review by Hedgemon4 — APPROVED** [review]

**Pull request merged**